src/main/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientFactory.java

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright (C) 2018 Expedia Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hotels.hcommon.hive.metastore.client.closeable;
+
+import java.lang.reflect.Proxy;
+
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import com.hotels.hcommon.hive.metastore.client.api.CloseableMetaStoreClient;
+import com.hotels.hcommon.hive.metastore.client.api.MetaStoreClientFactory;
+import com.hotels.hcommon.hive.metastore.compatibility.HiveMetaStoreClientCompatibility;
+import com.hotels.hcommon.hive.metastore.compatibility.HiveMetaStoreClientCompatibility12x;
+
+public final class CloseableMetaStoreClientFactory implements MetaStoreClientFactory {
+
+  private static final Logger log = LoggerFactory.getLogger(CloseableMetaStoreClientFactory.class);
+
+  private final IMetaStoreClient delegate;
+
+  private HiveMetaStoreClientCompatibility compatibility;
+
+  public CloseableMetaStoreClientFactory(IMetaStoreClient delegate) {
```

Comment:
this doesn't look right

Reply:
I've added a constructor that takes `Configuration` as an argument

In the ShuntingYard `CommonBeans` class you can do

```
@Bean
  Supplier<CloseableMetaStoreClient> replicaMetaStoreClientSupplier(
      HiveConf replicaHiveConf) {
    return new DefaultMetaStoreClientSupplier(new CloseableMetaStoreClientFactory(replicaHiveConf));
  }
```

And change `DefaultMetaStoreClientSupplier` to the following

```
public class DefaultMetaStoreClientSupplier implements Supplier<CloseableMetaStoreClient> {

  private final MetaStoreClientFactory metaStoreClientFactory;

  public DefaultMetaStoreClientSupplier(MetaStoreClientFactory metaStoreClientFactory) {
    this.metaStoreClientFactory = metaStoreClientFactory;
  }

  @Override
  public CloseableMetaStoreClient get() {
    return metaStoreClientFactory.newInstance();
  }

```

I preferred passing in Factory arguments via constructor cause then we can have a MetaStoreClientFactory interface